proposals/TP1-xds-transport-next.md

```diff
@@ -0,0 +1,1028 @@
+    // finalized, probably should be a RFC3986 pchar). This name allows
+    // reference via the #entry directive in ResourceLocator.
+    string name = 1 [(validate.rules).string.pattern = "^[0-9a-zA-Z_\\-\\.~:]+$"];
+
+    // The resource's logical version. It is illegal to have the same named xDS
+    // resource name at a given version with different resource payloads.
+    string version = 2;
+
+    // The resource payload, including type URL.
+    google.protobuf.Any resource = 3;
+  }
+
+  oneof resource_specifier {
+    option (validate.required) = true;
+
+    // A resource locator describing how the member resource is to be located.
+    ResourceLocator locator = 1;
+
+    // The resource is inlined in the list collection.
+    InlineEntry inline_entry = 2;
+  }
+}
+```
+
+A simple example of a listener collection fetch sequence:
+
+1. Client requests `xdstp://some-authority/envoy.config.listener.v3.ListenerCollection/foo.`
+2. Server responds with a list of resources: `[locator:
+   xdstp://some-authority/envoy.config.listener.v3.Listener/bar, locator:
+   xdstp://some-authority/envoy.config.listener.v3.Listener/baz].`
+3. Client requests both
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/bar `and
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/baz.`
+4. Server responds with resources contents for
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/bar `and
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/baz`.
+
+This involves two round-trips. Inlining of resources is supported, which can reduce RTT, i.e.
+
+1. Client requests `xdstp://some-authority/envoy.config.listener.v3.ListenerCollection/foo.`
+2. Server responds with a collection of literal resources inlined in `foo`:
+   `[inline_entry: xdstp://some-authority/envoy.config.listener.v3.Listener/bar,
+   inline_entry: xdstp://some-authority/envoy.config.listener.v3.Listener/baz].`
+
+
+#### Glob
+
+Having explicit collection resources allows delta xDS to provide delta updates
+on individual resources but the resource list itself can become a bottleneck for
+frequently updated large collections of resources, e.g. with 1M endpoints updating
+every 10s.  For this situation, we introduce *glob
+collections*, which allow the `<T>Collection` directory to be elided when
+resources in a collection share a directory structure and a `/*` component is appended to
+the resource path. Continuing the previous example, this now looks like:
+
+1. Client requests
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/foo/*.`
+2. Server responds with resources
+   [`xdstp://some-authority/envoy.config.listener.v3.Listener/foo/bar,
+   xdstp://some-authority/envoy.config.listener.v3.Listener/foo/baz].`
+
+Since each resource returned is subject to independent update via delta xDS and
+there is no explicit collection directory to update, glob collections are highly
+scalable.
```

Comment:
Consider the LEDS case.  We want the server to be able to do two things efficiently:

1. Send updates for individual endpoints that have changed without resending other endpoints that have not changed.
2. Dynamically change the *set* of endpoints for a given cluster -- i.e., to add and remove endpoints from the set without resending the whole set.

In list collections, there is one resource defining the set of resources and another resource defining the individual resources (assuming non-inlined entries in the collection).  So with a list collection, it's possible to do (1), but it's not possible to do (2) -- adding or removing an endpoint from the set would require resending the entire list collection resource.  Glob collections allow us to do (2), because we can simply send an add or a delete for individual endpoints that we want to add or remove from the set.

Reply:
The next section, `Use cases`, goes into this, but I've updated with an example clarifying the scalability argument.